src/stores/GuildMemberRoleStore.js

```diff
@@ -0,0 +1,192 @@
+      if (this._roles.includes(roleOrRoles.id)) return Promise.resolve(this.member);
+      return this.client.api.guilds(this.guild.id).members(this.member.id).roles(roleOrRoles.id)
+        .put({ reason })
+        .then(() => {
+          const clone = this.member._clone();
+          const roles = clone.roles._roles;
+          if (!roles.includes(roleOrRoles.id)) {
+            roles.push(roleOrRoles.id);
+            clone.roles._patch(roles);
+          }
+          return clone;
+        });
+    } else {
+      let allRoles = this._roles.slice();
+      for (let role of roleOrRoles instanceof Collection ? roleOrRoles.values() : roleOrRoles) {
+        role = this.resolve(role);
+        if (!role) {
+          return Promise.reject(new TypeError('INVALID_TYPE', 'roles',
+            'Array or Collection of Roles or Snowflakes', true));
+        }
+        if (!allRoles.includes(role.id)) allRoles.push(role.id);
+      }
+      return this.member.edit({ roles: allRoles }, reason);
+    }
+  }
+
+  /**
+   * Sets the roles applied to the member.
+   * @param {Collection<Snowflake, Role>|RoleResolvable[]} roles The roles or role IDs to apply
+   * @param {string} [reason] Reason for applying the roles
+   * @returns {Promise<GuildMember>}
+   * @example
+   * // Set the member's roles to a single role
+   * guildMember.roles.set(['391156570408615936'])
+   *   .then(console.log)
+   *   .catch(console.error);
+   * @example
+   * // Remove all the roles from a member
+   * guildMember.roles.set([])
+   *   .then(member => console.log(`Member roles is now of ${member.roles.size} size`))
+   *   .catch(console.error);
+   */
+  set(roles, reason) {
+    return this.member.edit({ roles }, reason);
+  }
+
+  /**
+   * Removes a role (or multiple roles) from the member
+   * @param {RoleResolvable|RoleResolvable[]|Collection<Snowflake, Role>} roleOrRoles The role or roles to remove
+   * @param {string} [reason] Reason for removing the role(s)
+   * @returns {Promise<GuildMember>}
+   */
+  remove(roleOrRoles, reason) {
+    if (this.resolve(roleOrRoles)) {
+      roleOrRoles = this.resolve(roleOrRoles);
+      if (!roleOrRoles) return Promise.reject(new TypeError('INVALID_TYPE', 'role', 'Role nor a Snowflake'));
+      if (!this._roles.includes(roleOrRoles.id)) return Promise.resolve(this.member);
+      return this.client.api.guilds(this.guild.id).members(this.member.id).roles(roleOrRoles.id)
+        .delete({ reason })
+        .then(() => {
+          const clone = this.member._clone();
+          const roles = clone.roles._roles;
+
+          const index = roles.indexOf(roleOrRoles.id);
+          if (~index) roles.splice(index, 1);
+
+          clone.roles._patch(roles);
+          return clone;
+        });
+    } else {
+      const allRoles = this._roles.slice();
+      for (let role of roleOrRoles instanceof Collection ? roleOrRoles.values() : roleOrRoles) {
```

Comment:
this whole loop is virtually identical to to that of the `add` method. hopefully removing the type conditional will help reduce this code duplication, but if it doesn't, please try to abstract this a bit.